### ISSUE TITLE
Gpu underutilization

### ISSUE THREAD

**Issue by ADAM-CT** [msr-fiddle/pipedream#29]
hello,This is my gpu utilization, I don't know which step caused the problem that led to the low utilization of gpu1
|   0  Tesla P100-SXM2...  Off  | 00000000:89:00.0 Off |                    0 |
| N/A   49C    P0   180W / 300W |   9481MiB / 16280MiB |    100%      Default |
+-------------------------------+----------------------+----------------------+
|   1  Tesla P100-SXM2...  Off  | 00000000:8A:00.0 Off |                    0 |
| N/A   35C    P0    44W / 300W |    657MiB / 16280MiB |      1%      Default |

----
Environment: 2 gpus in dgx-1
Bandwidth 20GB/s (considered NVLINK bandwidth)
Run the demo:
step1: CUDA_VISIBLE_DEVICES=0 python main.py -a vgg16 -b 32 --data_dir  ./data 

step2: python optimizer_graph_hierarchical.py -f ../profiler/image_classification/profiles/vgg16/graph.txt -n 2 --activation_compression_ratio 1 -o vgg16_partitioned --network_bandwidths [21474836480]

step3: python convert_graph_to_model.py -f vgg16_partitioned/gpus=2.txt -n VGG16Partitioned -a vgg16 -o ../runtime/image_classification/models/vgg16/gpus=2 --stage_to_num_ranks 0:1,1:1

step4: docker0 的ip为 10.1.1.4 ，docker1 的ip为 10.1.1.5
       
in docker0：
python main_with_runtime.py --module models.vgg16.gpus=2 -b 32 --data_dir  ./data  --rank 0 --local_rank 0 --master_addr 10.1.1.4 --config_path models/vgg16/gpus=2/hybrid_conf.json --distributed_backend gloo  --eval-batch-size 32

in docker1：
python main_with_runtime.py --module models.vgg16.gpus=2 -b 32 --data_dir  ./data  --rank 1 --local_rank 1 --master_addr 10.1.1.4 --config_path models/vgg16/gpus=2/hybrid_conf.json --distributed_backend gloo  --eval-batch-size 32

**Comment by deepakn94** [comment]
What does the partitioned model look like? How many layers per stage? What is the time-per-iteration compared to running everything on a single GPU?

**Comment by ADAM-CT** [comment]
The following is the model. After setting batch_size to 64, each iteration will take 0.445s
Epoch: 7 Step 14900     Learning rate: 0.100000
Epoch: [7][14900/20018] Time: 0.445 (0.445)     Epoch time [hr]: 1.842 (2.475)  Memory: 0.073 (0.094)   Loss: 6.9298 (6.9101)   Prec@1: 0.000 (0.101)   Prec@5: 1.562 (0.494)
Epoch: [7][14910/20018] Time: 0.443 (0.445)     Epoch time [hr]: 1.843 (2.475)  Memory: 0.073 (0.094)   Loss: 6.9141 (6.9101)   Prec@1: 0.000 (0.101)   Prec@5: 0.000 (0.494)
Epoch: [7][14920/20018] Time: 0.443 (0.445)     Epoch time [hr]: 1.845 (2.475)  Memory: 0.073 (0.094)   Loss: 6.8835 (6.9101)   Prec@1: 0.000 (0.101)   Prec@5: 0.000 (0.494)
Epoch: [7][14930/20018] Time: 0.445 (0.445)     Epoch time [hr]: 1.846 (2.475)  Memory: 0.073 (0.094)   

---
```
class Stage0(torch.nn.Module):
    def __init__(self):
          super(Stage0, self).__init__()
        self.layer2 = torch.nn.Conv2d(3, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer3 = torch.nn.ReLU(inplace=True)
        self.layer4 = torch.nn.Conv2d(64, 64, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer5 = torch.nn.ReLU(inplace=True)
        self.layer6 = torch.nn.MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        self.layer7 = torch.nn.Conv2d(64, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer8 = torch.nn.ReLU(inplace=True)
        self.layer9 = torch.nn.Conv2d(128, 128, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer10 = torch.nn.ReLU(inplace=True)
        self.layer11 = torch.nn.MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        self.layer12 = torch.nn.Conv2d(128, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer13 = torch.nn.ReLU(inplace=True)
        self.layer14 = torch.nn.Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer15 = torch.nn.ReLU(inplace=True)
        self.layer16 = torch.nn.Conv2d(256, 256, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer17 = torch.nn.ReLU(inplace=True)
        self.layer18 = torch.nn.MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        self.layer19 = torch.nn.Conv2d(256, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer20 = torch.nn.ReLU(inplace=True)
        self.layer21 = torch.nn.Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer22 = torch.nn.ReLU(inplace=True)
        self.layer23 = torch.nn.Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer24 = torch.nn.ReLU(inplace=True)
        self.layer25 = torch.nn.MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        self.layer26 = torch.nn.Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer27 = torch.nn.ReLU(inplace=True)
        self.layer28 = torch.nn.Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer29 = torch.nn.ReLU(inplace=True)
        self.layer30 = torch.nn.Conv2d(512, 512, kernel_size=(3, 3), stride=(1, 1), padding=(1, 1))
        self.layer31 = torch.nn.ReLU(inplace=True)
        self.layer32 = torch.nn.MaxPool2d(kernel_size=2, stride=2, padding=0, dilation=1, ceil_mode=False)
        self.layer35 = torch.nn.Linear(in_features=25088, out_features=4096, bias=True)
        self.layer36 = torch.nn.ReLU(inplace=True)
        self.layer37 = torch.nn.Dropout(p=0.5)
        self.layer38 = torch.nn.Linear(in_features=4096, out_features=4096, bias=True)
        self._initialize_weights()


     def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m, torch.nn.Conv2d):
                torch.nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
                if m.bias is not None:
                    torch.nn.init.constant_(m.bias, 0)
            elif isinstance(m, torch.nn.BatchNorm2d):
                torch.nn.init.constant_(m.weight, 1)
                torch.nn.init.constant_(m.bias, 0)
            elif isinstance(m, torch.nn.Linear):
                torch.nn.init.normal_(m.weight, 0, 0.01)
                torch.nn.init.constant_(m.bias, 0)

    def forward(self, input0):
        out0 = input0.clone()
        out2 = self.layer2(out0)
        out3 = self.layer3(out2)
        out4 = self.layer4(out3)
        out5 = self.layer5(out4)
        out6 = self.layer6(out5)
        out7 = self.layer7(out6)
        out8 = self.layer8(out7)
        out9 = self.layer9(out8)
        out10 = self.layer10(out9)
        out11 = self.layer11(out10)
        out12 = self.layer12(out11)
        out13 = self.layer13(out12)
        out14 = self.layer14(out13)
        out15 = self.layer15(out14)
        out16 = self.layer16(out15)
        out17 = self.layer17(out16)
        out18 = self.layer18(out17)
        out19 = self.layer19(out18)
        out20 = self.layer20(out19)
        out21 = self.layer21(out20)
        out22 = self.layer22(out21)
        out23 = self.layer23(out22)
        out24 = self.layer24(out23)
        out25 = self.layer25(out24)
        out26 = self.layer26(out25)
        out27 = self.layer27(out26)
        out28 = self.layer28(out27)
        out29 = self.layer29(out28)
        out30 = self.layer30(out29)
        out31 = self.layer31(out30)
        out32 = self.layer32(out31)
        out33 = out32.size(0)
        out34 = out32.view(out33, -1)
        out35 = self.layer35(out34)
        out36 = self.layer36(out35)
        out37 = self.layer37(out36)
        out38 = self.layer38(out37)
        return out38

```
---
```
class Stage1(torch.nn.Module):
    def __init__(self):
        super(Stage1, self).__init__()
        self.layer1 = torch.nn.ReLU(inplace=True)
        self.layer2 = torch.nn.Dropout(p=0.5)
        self.layer3 = torch.nn.Linear(in_features=4096, out_features=1000, bias=True)
        self._initialize_weights()

    def _initialize_weights(self):
        for m in self.modules():
            if isinstance(m, torch.nn.Conv2d):
                torch.nn.init.kaiming_normal_(m.weight, mode='fan_out', nonlinearity='relu')
                if m.bias is not None:
                    torch.nn.init.constant_(m.bias, 0)
            elif isinstance(m, torch.nn.BatchNorm2d):
                torch.nn.init.constant_(m.weight, 1)
                torch.nn.init.constant_(m.bias, 0)
            elif isinstance(m, torch.nn.Linear):
                torch.nn.init.normal_(m.weight, 0, 0.01)
                torch.nn.init.constant_(m.bias, 0)

    def forward(self, input0):
        out0 = input0.clone()
        out1 = self.layer1(out0)
        out2 = self.layer2(out1)
        out3 = self.layer3(out2)
        return out3


```

**Comment by ADAM-CT** [comment]
Run step1:
step1: CUDA_VISIBLE_DEVICES=0 python main.py -a vgg16 -b 32 --data_dir ./data
The output of step1 is:
      Total accounted time: 467.428 ms
The single gpu cost 467ms,PipeDream takes an average of 445ms per iteration, not much better than a single gpu.So I wonder if it is the problem of the bandwidth setting in step2?It should not be set to [21474836480].But what should be the bandwidth between the same server?

**Comment by ADAM-CT** [comment]
When I use the default bandwidth: [1000000000](0.93GB/s), pipeDream only needs 363ms for average iteration. I want to know how do you determine the bandwidth of multiple gpus on a server?My environment is DGX-1 2 gpus.

**Comment by ADAM-CT** [comment]
Dp: a epoch only takes 1.5 hours
I tested the optimal partition of PipeDream as: stage0:(0,10). Stage1 :(10,40), 2 hours for a epoch。
I wonder if DP is better than PipeDream with fewer gpus?

**Comment by deepakn94** [comment]
Okay: so the underutilization in the first example makes sense. The provided bandwidth is really high, so the optimizer thinks any partitioning is good, and puts most of the model's computation in the first stage. This leads to compute imbalance across the two stages, and the second GPU being severely underutilized.

Regarding the DP vs. PipeDream time-per-epochs: for this particular model (VGG-16), it's probably hard to find a model split that balances compute times evenly across the two stages. If you increase the total number of GPUs, you will probably see the stage replication factor of the first stage increase in order to better balance the compute times (and also leverage the fact that all-reduce communication for convolutional layers is cheap). It should be possible to find good 2-GPU configurations for PipeDream for other models (especially models with the same layer repeated multiple times).

**Comment by deepakn94** [comment]
Going to close this. Feel free to re-open if the above doesn't answer your question.
